Gems/Atom/RHI/Vulkan/Code/Source/RHI/RenderPass.cpp

```diff
@@ -16,6 +16,356 @@ namespace AZ
 {
     namespace Vulkan
     {
+        namespace Internal
+        {
+            enum class AttachmentType : uint32_t
+            {
+                Color, // Color render target attachment
+                DepthStencil, // Depth stencil attachment
+                Resolve, // Resolve attachment
+                InputAttachment, // An input attachment that is the output of a previous subpass.
+                Preserve, // An attachment that is not being used by the subpass but that it will be preserved.
+                ShadingRate, // An attachment that is used for specifiying shading rate.
+                Count
+            };
+            static const uint32_t AttachmentTypeCount = static_cast<uint32_t>(AttachmentType::Count);
+
+            using RenderPassResult = AZStd::tuple<VkResult, VkRenderPass>;
+
+            template<
+                typename VkRenderPassCreateInfoTraits,
+                typename VkAttachmentDescriptionTraits,
+                typename VkAttachmentReferenceTraits,
+                typename VkSubpassDescriptionTraits,
+                typename VkSubpassDependencyTraits>
+            struct VkRenderpassBuilder
+            {
+            public:
+
+                VkRenderpassBuilder(const Device& device)
+                    : m_device(device)
+                {}
+
+                using VkRenderPassCreateInfoType = typename VkRenderPassCreateInfoTraits::value_type;
+                using VkAttachmentDescriptionType = typename VkAttachmentDescriptionTraits::value_type;
+                using VkAttachmentReferenceType = typename VkAttachmentReferenceTraits::value_type;
+                using VkSubpassDescriptionType = typename VkSubpassDescriptionTraits::value_type;
+                using VkSubpassDependencyType = typename VkSubpassDependencyTraits::value_type;
+
+                struct SubpassInfo
+                {
+                    AZStd::array<AZStd::vector<VkAttachmentReferenceType>, AttachmentTypeCount> m_attachmentReferences;
+                    AZStd::vector<uint32_t> m_preserveAttachments;
+                    VkFragmentShadingRateAttachmentInfoKHR m_shadingRateAttachmentExtension;
+                };
+
+                RenderPassResult Build(const RenderPass::Descriptor& descriptor)
```

Comment:
Would be good to add a comment above each method since they dont exist in the header file